platform-sdk/docs/proposals/prehandle-buffer/prehandle-buffer-proposal.md

```diff
@@ -0,0 +1,183 @@
+
+- `minSequence` - the minimum sequence number that has not been pre-handled.
+- `startIndex` - the array index of the minimum sequence number.
+
+| Min sequence | Start index |
+|--------------|-------------|
+| 100          | 3           |
+
+Checking if an element has been pre-handled is:
+- if its smaller than `minSequence`, it has been pre-handled.
+- if its larger than `minSequence + arraySize`, it has not been pre-handled.
+- if its between `minSequence` and `minSequence + arraySize`, we can check the array.
+
+Updating the data structure:
+Suppose our input is that the event with the sequence number 100. We would first update the array:
+
+| Array index  | 0   | 1   | 2   | 3   | 4   | 5   |
+|--------------|-----|-----|-----|-----|-----|-----|
+| Sequence     | 103 | 104 | 105 | 100 | 101 | 102 |
+| Pre-handled? | N   | Y   | Y   | Y   | Y   | Y   |
+
+Our next step would be to check if we can shift the data. We start from `startIndex` and iterate until we find an event
+which has not been handled. Then we update the values appropriately:
+
+| Array index  | 0   | 1   | 2   | 3   | 4   | 5   |
+|--------------|-----|-----|-----|-----|-----|-----|
+| Sequence     | 103 | 104 | 105 | 106 | 107 | 108 |
+| Pre-handled? | N   | Y   | Y   | N   | N   | N   |
+
+| Min sequence | Start index |
+|--------------|-------------|
+| 103          | 0           |
+
+#### Wiring changes
+
+The following changes would be made to the wiring:
+- A new component, `PreHandleBuffer`, would be added to the wiring.
+- The `PcesSequencer` would be renamed to `EventSequencer` since it is no longer only used for PCES events.
+- The `TransactionPrehandler` would be wired to the `EventSequencer` instead of the `OrphanBuffer` because we need to
+  ensure that events have a sequence number before they are pre-handled. This is necessary because the `PreHandleBuffer`
+  will track events based on their sequence number.
+- The `TransactionPrehandler` would output pre-handled events to the `PreHandleBuffer`. This will notify the buffer 
+  which events have been pre-handled.
+- The `PreHandleBuffer` would be wired in-between the `ConsensusEngine` and the `RoundDurabilityBuffer`. It will buffer
+  rounds until pre-handle is complete.
+
+### Metrics
+
+- the metric `WAITING_FOR_PREHANDLE` phase of the `RoundHandlingMetrics` should be removed.
+- the wiring should track the number of tasks waiting on the new component.
+- the new component should have a metric that tracks the number of rounds that are waiting for pre-handle to complete.
+- we should also track the capacity of the circular buffer, just to be sure that it is not growing to much.
+
+### Performance
+
+No performance impacts are expected. There could be a potential for a performance improvement in the case where events 
+are not pre-handled by the time they reach consensus. In practice, this is not a common occurrence. The only situation 
+where this is expected might be a low latency network.
```

Comment:
How much memory would the status array occupy?

How much memory would the buffered rounds occupy?

Could the extra memory usage affect the performance of other components in platform or services?

What about a case where an event fails to complete pre-handling and is never marked as pre-handled. Would those array and the buffer grow until OOM happens then?

Reply:
> How much memory would the status array occupy?

I expect very little, like max 1000 elements

> How much memory would the buffered rounds occupy?

Not any more than they do now, they will just be buffered in a different place.

> Could the extra memory usage affect the performance of other components in platform or services?

I don't expect any extra memory usage

> What about a case where an event fails to complete pre-handling and is never marked as pre-handled. Would those array and the buffer grow until OOM happens then?

Probably. Right now, if that happens, the node will just halt forever. We could add the number of buffered rounds to the health monitor, it would prevent the node from OOMing.

Reply:
Thanks for the details. If we choose to go with this solution, then I suggest to describe these memory-related issues in the Performance section and add the health monitor-related modifications into the design document to ensure we implement this protection from OOMs.